output/sarif.go

```diff
@@ -0,0 +1,496 @@
+	result := Result{
+		Message: successResultType.description,
+	}
+	return processResults(run, []Result{result}, successResultType, fileName, namespace, ruleMap)
+}
+
+// createSkippedResult creates a skipped result for a given file and namespace
+func createSkippedResult(run *sarifRun, fileName, namespace string, ruleMap map[string]bool) error {
+	result := Result{
+		Message: skippedResultType.description,
+	}
+	return processResults(run, []Result{result}, skippedResultType, fileName, namespace, ruleMap)
+}
+
+// Output outputs the results in SARIF format.
+func (s *SARIF) Output(results []CheckResult) error {
+	startTime := time.Now().UTC()
+
+	// Create SARIF report structure
+	report := sarifReport{
+		Schema:  sarifSchemaURI,
+		Version: sarifVersion,
+		Runs: []sarifRun{
+			{
+				Tool: sarifTool{
+					Driver: sarifDriver{
+						Name:           toolName,
+						InformationURI: toolURI,
+						Rules:          []sarifRule{},
+					},
+				},
+				Results:     []sarifResult{},
+				Invocations: []sarifInvocation{},
+			},
+		},
+	}
+
+	run := &report.Runs[0]
+	ruleMap := make(map[string]bool)
+
+	// Process all results
+	for _, result := range results {
+		err := processResults(run, result.Failures, failureResultType, result.FileName, result.Namespace, ruleMap)
+		if err != nil {
+			return fmt.Errorf("process failures: %w", err)
+		}
+		err = processResults(run, result.Warnings, warningResultType, result.FileName, result.Namespace, ruleMap)
+		if err != nil {
+			return fmt.Errorf("process warnings: %w", err)
+		}
+		err = processResults(run, result.Exceptions, exceptionResultType, result.FileName, result.Namespace, ruleMap)
+		if err != nil {
+			return fmt.Errorf("process exceptions: %w", err)
+		}
+
+		// Add success result if no failures/warnings/exceptions
+		if len(result.Failures) == 0 && len(result.Warnings) == 0 && len(result.Exceptions) == 0 {
+			if result.Successes > 0 {
+				err = createSuccessResult(run, result.FileName, result.Namespace, ruleMap)
+				if err != nil {
+					return fmt.Errorf("create success result: %w", err)
+				}
+			} else {
+				err = createSkippedResult(run, result.FileName, result.Namespace, ruleMap)
+				if err != nil {
+					return fmt.Errorf("create skipped result: %w", err)
+				}
+			}
+		}
+	}
+
+	// Add invocation information
+	exitCode := 0
+	exitDesc := exitNoViolations
+	if hasFailures(results) {
+		exitCode = 1
+		exitDesc = exitViolations
+	} else if hasWarnings(results) {
+		exitCode = 0
+		exitDesc = exitWarnings
+	} else if hasExceptions(results) {
+		exitCode = 0
+		exitDesc = exitExceptions
+	}
+
+	run.Invocations = []sarifInvocation{
+		{
+			ExecutionSuccessful: true,
+			ExitCode:            exitCode,
+			ExitCodeDescription: exitDesc,
+			StartTimeUtc:        startTime.Format(time.RFC3339),
+			EndTimeUtc:          time.Now().UTC().Format(time.RFC3339),
+		},
+	}
+
+	// Marshal to JSON
+	encoder := json.NewEncoder(s.writer)
+	encoder.SetIndent("", "  ")
```

Comment:
Is there a need for this? IIUC, the SARIF format is intended to be consumed by machines.

Reply:
I guess we can offload that to `jq` for development purposes. `go-sarif` has separate methods for this as in `Write()` and `PrettyWrite()`, so lets write it in machine-friendly format without whitespaces by default.